trie/hashtrie.go

```diff
@@ -121,3 +124,173 @@ func (t *HashTrie) Hash() common.Hash {
+		// The extension and the key are identical on the length of
+		// the extension, so st.stack[level].ext.Val should be a fullNode and
+		// the difference should be found there. Panic if this is
+		// not the case.
+		fn := st.stack[level].ext.Val.(*fullNode)
+
+		// The correct entry is the only one that isn't nil
+		for i := 15; i >= 0; i-- {
+			if fn.Children[i] != nil {
+				switch fn.Children[i].(type) {
+				// Only hash entries that are not already hashed
+				case *fullNode, *shortNode:
+					fn.Children[i], _ = st.hasher.hash(fn.Children[i], false)
+					st.top = level
+				default:
+				}
+				break
+			}
+		}
+
+		// That fullNode should have at most one non-hashNode child,
+		// hash it because no more nodes will be inserted in it.
+		if len(st.stack) == st.top+1 {
+			st.stack = append(st.stack, StackTrieItem{})
+		}
+
+		st.top++
+		keyUntilHere := len(st.stack[level].keyUntilHere) + len(st.stack[level].ext.Key) + 1
+		st.stack[level].branch.Children[key[keyUntilHere]] = &st.stack[st.top].ext
+		st.stack[st.top].keyUntilHere = key[:keyUntilHere]
+		st.stack[st.top].ext.Key = key[keyUntilHere:]
+		st.stack[st.top].ext.Val = hv
+		st.stack[st.top].ext.flags = nodeFlag{dirty: true}
+		st.stack[st.top].depth = st.stack[level].depth + 1
+	} else {
+		// extension keys differ, need to create a split and
+		// hash the former node.
+		whereitdiffers := 0
+		offset := len(st.stack[level].keyUntilHere)
+		for i := range st.stack[level].ext.Key {
+			if key[offset+i] != st.stack[level].ext.Key[i] {
+				whereitdiffers = i
+				break
+			}
+		}
+
+		// Start by hashing the node right after the extension,
+		// to free some space.
+		var hn node
+		switch st.stack[level].ext.Val.(type) {
+		case *fullNode:
+			h, _ := st.hasher.hash(st.stack[level].ext.Val, false)
+			hn = h.(hashNode)
+		case hashNode, valueNode:
+			hn = st.stack[level].ext.Val
+		default:
+			panic("Encountered unexpected node type")
+		}
+
+		// Allocate the next full node, it's going to be
+		// reused several times.
+		if len(st.stack) == st.top+1 {
+			st.stack = append(st.stack, StackTrieItem{})
+		}
+		st.top++
+
+		// Store the partially-hashed old node in the newly allocated
+		// slot, in order to finish the hashing.
+		slot := st.stack[level].ext.Key[whereitdiffers]
+		st.stack[st.top].ext.Key = st.stack[level].ext.Key[whereitdiffers+1:]
+		st.stack[st.top].ext.Val = hn
+		st.stack[st.top].ext.flags = nodeFlag{dirty: true}
+
+		// Hasher directement la branche si l'ext est vide
+		h, _ := st.hasher.hash(&st.stack[st.top].ext, false)
+		st.stack[level].branch.Children[slot] = h.(hashNode)
+		st.stack[level].ext.Val = &st.stack[level].branch
+		st.stack[level].ext.Key = st.stack[level].ext.Key[:whereitdiffers]
+
+		// Now use the newly allocated+hashed stack st.stack[level] to store
+		// the rest of the inserted (key, value) pair.
+		slot = key[whereitdiffers+len(st.stack[level].keyUntilHere)]
+		st.stack[level].branch.Children[slot] = &st.stack[st.top].ext
```

Comment:
Just realized that there is a potential bug here in the case two `fullNode`s are stored back to back.

Reply:
I'll merge it for now anyway